src/types/Writable.ts

```diff
@@ -32,5 +38,20 @@ export type Writable<T> = T extends WritablePrimitive
 			: T extends Set<infer X>
 				? Set<Writable<X>>
 				: T extends Object
-					? { [K in WritableKeys<T>]: Writable<T[K]> }
+					? {
+							// Capture required keys
+							[K in keyof T & RequiredKeys<T> & WritableKeys<T>]: Writable<
+								T[K]
+							>;
+						} & {
+							// Capture optional keys
+							[K in keyof T & OptionalKeys<T> & WritableKeys<T>]?: Writable<
+								T[K]
+							>;
```

Comment:
The crucial difference here being the `?` after the `[K in keyof T...]`, leaving the optional keys optional, and the required keys required.